blueoil/datasets/dataset_iterator_multi.py

```diff
@@ -0,0 +1,272 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Blueoil Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import os
+import queue
+import threading
+from multiprocessing import Pool
+
+import numpy as np
+import tensorflow as tf
+
+from blueoil.datasets.tfds import TFDSMixin
+
+_dataset = None
+
+
+def _prefetch_setup(dataset, seed, do_shuffle):
```

Comment:
@yd8534976 
Thanks. It's nice PR.
`dataset_iterator_multi` and `dataset_iterator` contains same code.
For example, `_apply_augmentations ` and `_xorshift32 ` are looks same.
How about to create `DatasetIteratorUtil` class to put theses same code?

Reply:
@iizukak 
Thanks for comments. Although both `dataset_iterator_multi ` and `dataset_iterator` contains `_apply_augmentations ` and `_concat_data`. They are different.

For example, in `dataset_iterator`
```
def _apply_augmentations(dataset, image, label):
    augmentor = dataset.augmentor
    pre_processor = dataset.pre_processor

    sample = {'image': image}

    if issubclass(dataset.__class__, SegmentationBase):
        sample['mask'] = label
    elif issubclass(dataset.__class__, ObjectDetectionBase):
        sample['gt_boxes'] = label
    elif issubclass(dataset.__class__, KeypointDetectionBase):
        sample['joints'] = label
    else:
        sample['label'] = label

    if callable(augmentor) and dataset.subset == 'train':
        sample = augmentor(**sample)

    if callable(pre_processor):
        sample = pre_processor(**sample)

    image = sample['image']

    if issubclass(dataset.__class__, SegmentationBase):
        label = sample['mask']
    elif issubclass(dataset.__class__, ObjectDetectionBase):
        label = sample['gt_boxes']
    elif issubclass(dataset.__class__, KeypointDetectionBase):
        label = sample['heatmap']
    else:
        label = sample['label']

    # FIXME(tokunaga): dataset should not have their own data format
    if dataset.data_format == "NCHW":
        image = np.transpose(image, [2, 0, 1])

    return (image, label)
```
In `dataset_iterator_multi`
```
def _apply_augmentations(dataset, sample):
    augmentor = dataset.augmentor
    pre_processor = dataset.pre_processor
    if callable(augmentor) and dataset.subset == 'train':
        sample = augmentor(**sample)
    if callable(pre_processor):
        sample = pre_processor(**sample)
    # FIXME(tokunaga): dataset should not have their own data format
    if dataset.data_format == "NCHW":
        sample['image'] = np.transpose(sample['image'], [2, 0, 1])
    return sample
```

Reply:
I see.
How about `_xorshift32`.
All functions in `dataset_iterator_multi.py` are different from original one?
If so, there are no problem.

Reply:
@iizukak  Not all. but most of them are different. `_xorshift32` is the same as you said. I'm not sure is it worth doing this? 🤔

Reply:
@yd8534976 
I'm also not sure.
Please wait for readability reviewer.